tsl/src/data_node.c

```diff
@@ -899,6 +902,20 @@ data_node_attach(PG_FUNCTION_ARGS)
 		}
 	}
 
+	/*
+	 * Change to the hypertable owner so that the same permissions will be set up on the
+	 * datanode being attached to as well. We need to do this explicitly because the
+	 * caller of this function could be a superuser and we definitely don't want to create
+	 * this hypertable with superuser ownership on the datanode being attached to!
+	 */
+	rel = table_open(ht->main_table_relid, AccessShareLock);
+	uid = rel->rd_rel->relowner;
+	table_close(rel, AccessShareLock);
```

Comment:
Maybe we want to keep this lock until the end of the transaction?

I am thinking of the case where there's a concurrent change of owner and maybe there's a race condition. For example, we release the lock here, then another process changes the owner (which only applies to the nodes attached before this command completes), then this code proceeds with creating the hypertable on the data node with the old owner.



```suggestion
	table_close(rel, NoLock);
```

Reply:
Maybe it is even possible to create an isolation test for that case?

Reply:
Maybe the lock type should also be `AccessExclusiveLock` to block other operations while we attach the node, e.g., owner change.

Reply:
The `owner change` takes an `AccessExclusiveLock` so it should be ok enough to close the table with `NoLock` here. Done with that change.

Reply:
Ok, that seems reasonable.